docs/datatypes/tsemaphore.md

```diff
@@ -0,0 +1,99 @@
+---
+id: datatypes_tsemaphore
+title:  "TSemaphore"
+---
+`TSemaphore` is a semaphore with transactional semantics that can be used to control access to a common resource. It 
+holds a certain number of permits, and permits may be acquired or released.
+
+## Create a TSemaphore
+Creating a `TSemaphore` with 10 permits:
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tSemaphoreCreate: STM[Nothing, TSemaphore] = TSemaphore.make(10L)
+```
+
+## Acquire a permit
+Acquiring a permit reduces the number of remaining permits that the `TSemaphore` contains. Acquiring a permit is done 
+when a user wants to access a common shared resource:
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tSemaphoreAcq: UIO[TSemaphore] = (for {
+  tSem <- TSemaphore.make(2L)
+  _    <- tSem.acquire
+} yield tSem).commit
+```
+Note that if you try to acquire a permit when there are no more remaining permits in the semaphore then execution will be blocked until a permit is ready to be acquired.
+
+## Release a permit
+Once you have finished accessing the shared resource, you must release your permit so other parties can access the 
+shared resource:
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tSemaphoreRelease: UIO[TSemaphore] = (for {
+  tSem <- TSemaphore.make(1L)
+  _    <- tSem.acquire
+  _    <- tSem.release
+} yield tSem).commit
+```
+## Retrieve available permits
+You can query for the remaining amount of permits in the TSemaphore by using `available`:
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+val tSemaphoreAvailable: UIO[Long] = (for {
+  tSem <- TSemaphore.make(2L)
+  _    <- tSem.acquire
+  cap  <- tSem.available
+} yield cap).commit
+```
+The above code creates a TSemaphore with two permits and acquires one permit without releasing it. Here, `available`
+will report that there is a single permit left.
+
+## Execute an arbitrary STM action with automatic acquire and release
+You can choose to execute any arbitrary STM action that requires acquiring and releasing permit on TSemaphore as part
+of the same transaction. Rather than doing:
+```scala mdoc:silent
+import zio._
+import zio.stm._
+
+def yourSTMAction: STM[Nothing, Unit] = STM.unit
+
+val tSemaphoreWithoutPermit: IO[Nothing, Unit] = 
+  (for {
+    sem <- TSemaphore.make(1L)
+    _   <- sem.acquire
+    a   <- yourSTMAction
+    _   <- sem.release
+  } yield a).commit
+```
+You can simply use `withPermit` instead:
```

Comment:
Maybe emphasize more that outside of advanced use cases this is generally best practice because it makes sure that we release the permits when we are done.

Reply:
done :-), have added a note at the end of the code sample